grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/GrpcDocStringExtractor.java

```diff
@@ -115,11 +115,44 @@ protected boolean acceptFile(String filename) {
                          .collect(toImmutableMap(Entry::getKey, Entry::getValue));
     }
 
+    // A path can be documented if it is not an option.
+    // First, the option documentations are not supported in DocService yet.
+    // Second, the option documentations cause duplicate keys to appear while parsing the file
+    // which makes DocService fail to initialize.
+    private static boolean canDocumentPath(List<Integer> path) {
+        switch (path.get(0)) {
+            case FileDescriptorProto.MESSAGE_TYPE_FIELD_NUMBER:
+                return true;
+            case FileDescriptorProto.ENUM_TYPE_FIELD_NUMBER:
+                // If this is an enum option, skip it.
+                if (path.size() > 2 && path.get(2) == EnumDescriptorProto.OPTIONS_FIELD_NUMBER) {
+                    return false;
+                }
+                return true;
+            case FileDescriptorProto.SERVICE_FIELD_NUMBER:
+                if (path.size() > 2) {
+                    // If there is a fifth path, it means the target is an `option` and we can skip it.
+                    if (path.size() > 4) {
+                        return false;
+                    }
+
+                    return path.get(2) != ServiceDescriptorProto.OPTIONS_FIELD_NUMBER;
+                }
+                return true;
+            default:
+                return false;
+        }
+    }
+
     // A path is field number and indices within a list of types, going through a tree of protobuf
     // descriptors. For example, the 2nd field of the 3rd nested message in the 1st message in a file
     // would have path [MESSAGE_TYPE_FIELD_NUMBER, 0, NESTED_TYPE_FIELD_NUMBER, 2, FIELD_FIELD_NUMBER, 1]
     @Nullable
     private static String getFullName(FileDescriptorProto descriptor, List<Integer> path) {
+        if (!canDocumentPath(path)) {
```

Comment:
Instead of filtering before entering the method, what do you think of checking while traversing each path?

e.g.
https://github.com/line/armeria/compare/master...jrhee17:suggestion/fix-grpc-docs-options?expand=1

Reply:
My initial intention was short-circuiting but probably pulling logic into append methods is easier to read & maintain.